File: Day-13/merkle-tree-sample/java/MerkleTree.java
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;

public class MerkleTree {
    public static final class ProofStep {
        private final String siblingHash;
        private final boolean siblingOnLeft;

        private ProofStep(String siblingHash, boolean siblingOnLeft) {
            this.siblingHash = siblingHash;
            this.siblingOnLeft = siblingOnLeft;
        }

        public String getSiblingHash() {
            return siblingHash;
        }

        public boolean isSiblingOnLeft() {
            return siblingOnLeft;
        }
    }

    private final List<String> values;
    private final List<List<String>> levels;

    public MerkleTree(List<String> values) {
        if (values == null || values.isEmpty()) {
            throw new IllegalArgumentException("values must be non-empty");
        }
        for (String value : values) {
            if (value == null || value.isBlank()) {
                throw new IllegalArgumentException("values must not contain empty items");
            }
        }

        this.values = new ArrayList<>(values);
        this.levels = buildLevels(values);
    }

    public String getRootHash() {
        return levels.get(levels.size() - 1).get(0);
    }

    public List<ProofStep> proofFor(int index) {
        if (index < 0 || index >= values.size()) {
            throw new IndexOutOfBoundsException("index out of range");
        }

        List<ProofStep> proof = new ArrayList<>();
        int currentIndex = index;

        for (int levelIndex = 0; levelIndex < levels.size() - 1; levelIndex += 1) {
            List<String> level = levels.get(levelIndex);
            int siblingIndex = currentIndex % 2 == 0 ? currentIndex + 1 : currentIndex - 1;
            if (siblingIndex >= level.size()) {
                siblingIndex = currentIndex;
            }

            proof.add(new ProofStep(level.get(siblingIndex), siblingIndex < currentIndex));
            currentIndex /= 2;
        }

        return proof;
    }

    public static boolean verifyProof(String value, List<ProofStep> proof, String expectedRoot) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException("value must be non-empty");
        }

        String currentHash = leafHash(value);
        for (ProofStep step : proof) {
            if (step.isSiblingOnLeft()) {
                currentHash = parentHash(step.getSiblingHash(), currentHash);
            } else {
                currentHash = parentHash(currentHash, step.getSiblingHash());
            }
        }
        return currentHash.equals(expectedRoot);
    }

    private static List<List<String>> buildLevels(List<String> values) {
        List<List<String>> levels = new ArrayList<>();
        List<String> leaves = new ArrayList<>();
        for (String value : values) {
            leaves.add(leafHash(value));
        }
        levels.add(leaves);

        while (levels.get(levels.size() - 1).size() > 1) {
            List<String> current = levels.get(levels.size() - 1);
            List<String> nextLevel = new ArrayList<>();

            for (int index = 0; index < current.size(); index += 2) {
                String left = current.get(index);
                String right = index + 1 < current.size() ? current.get(index + 1) : current.get(index);
                nextLevel.add(parentHash(left, right));
            }

            levels.add(nextLevel);
        }

        return levels;
    }

    private static String leafHash(String value) {
        return sha256Hex("leaf:" + value);
    }

    private static String parentHash(String left, String right) {
        return sha256Hex("node:" + left + ":" + right);
    }

    private static String sha256Hex(String value) {
        try {
            byte[] digest = MessageDigest.getInstance("SHA-256").digest(value.getBytes(StandardCharsets.UTF_8));
            StringBuilder builder = new StringBuilder(digest.length * 2);
            for (byte current : digest) {
                builder.append(String.format("%02x", current));
            }
            return builder.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 algorithm unavailable", e);
        }
    }
}

File: Day-13/merkle-tree-sample/python/merkle_tree.py
from __future__ import annotations

import hashlib
from dataclasses import dataclass


@dataclass(frozen=True)
class ProofStep:
    sibling_hash: str
    sibling_on_left: bool


class MerkleTree:
    def __init__(self, values: list[str]) -> None:
        if not values:
            raise ValueError("values must be non-empty")
        if any(not value for value in values):
            raise ValueError("values must not contain empty items")

        self._values = values[:]
        self._levels = self._build_levels(values)

    @property
    def root_hash(self) -> str:
        return self._levels[-1][0]

    def proof_for(self, index: int) -> list[ProofStep]:
        if index < 0 or index >= len(self._values):
            raise IndexError("index out of range")

        proof: list[ProofStep] = []
        current_index = index

        for level in self._levels[:-1]:
            sibling_index = current_index - 1 if current_index % 2 else current_index + 1
            if sibling_index >= len(level):
                sibling_index = current_index

            proof.append(
                ProofStep(
                    sibling_hash=level[sibling_index],
                    sibling_on_left=sibling_index < current_index,
                )
            )
            current_index //= 2

        return proof

    @staticmethod
    def verify_proof(value: str, proof: list[ProofStep], expected_root: str) -> bool:
        if not value:
            raise ValueError("value must be non-empty")

        current_hash = MerkleTree._leaf_hash(value)
        for step in proof:
            if step.sibling_on_left:
                current_hash = MerkleTree._parent_hash(step.sibling_hash, current_hash)
            else:
                current_hash = MerkleTree._parent_hash(current_hash, step.sibling_hash)
        return current_hash == expected_root

    @classmethod
    def _build_levels(cls, values: list[str]) -> list[list[str]]:
        levels: list[list[str]] = [[cls._leaf_hash(value) for value in values]]

        while len(levels[-1]) > 1:
            current = levels[-1]
            next_level: list[str] = []
            for index in range(0, len(current), 2):
                left = current[index]
                right = current[index + 1] if index + 1 < len(current) else current[index]
                next_level.append(cls._parent_hash(left, right))
            levels.append(next_level)

        return levels

    @staticmethod
    def _leaf_hash(value: str) -> str:
        return hashlib.sha256(f"leaf:{value}".encode("utf-8")).hexdigest()

    @staticmethod
    def _parent_hash(left: str, right: str) -> str:
        return hashlib.sha256(f"node:{left}:{right}".encode("utf-8")).hexdigest()
